pppd/sys-linux.c

```diff
@@ -1491,6 +1494,118 @@ get_ppp_stats_ioctl(int u, struct pppd_stats *stats)
     return 1;
 }
 
+/********************************************************************
+ * get_ppp_stats_rtnetlink - return statistics for the link, using rtnetlink
+ * This provides native 64-bit counters.
+ */
+#ifdef RTM_GETSTATS
+static int
+get_ppp_stats_rtnetlink(int u, struct pppd_stats *stats)
+{
+    static int rtnl_fd = -1;
+
+    struct sockaddr_nl nladdr;
+    struct {
+        struct nlmsghdr nlh;
+        struct if_stats_msg ifsm;
+    } nlreq;
+    struct {
+        struct nlmsghdr nlh;
+	union {
+	    struct nlmsgerr nlerr;
+	    struct {
+		struct rtmsg rth;
+		struct rtnl_link_stats64 stats;
+	    };
+	};
+    } nlresp;
+    ssize_t nlresplen;
+    struct iovec iov;
+    struct msghdr msg;
+
+    if (rtnl_fd < 0) {
+	rtnl_fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	if (rtnl_fd < 0) {
+	    error("get_ppp_stats_rtnetlink: error creating NETLINK socket: %m (line %d)", __LINE__);
+	    return 0;
+	}
+
+	memset(&nladdr, 0, sizeof(nladdr));
+	nladdr.nl_family = AF_NETLINK;
+	nladdr.nl_pid = getpid();
```

Comment:
I noticed in your code.  My testing last night disagreed, but I must've screwed something else up as it works as you state now on my test code block.  Adjusted locally so long.

BTW, reading other examples and stuff, pppd here seems to be makinga few (IMHO reasonable) assumptions about the layout of NL messages.  However, should a compiler decide to pack the structures this will break (eg, from top of memory, struct rtmsg is 12 bytes, but the compiler adds 4 bytes padding to get to the rtnl_link_stats64 structure, on amd64 at least).

There are MACROs in linux/netlink.h and linux/rtnetlink.h to "get to these things the 'correct' way" but my experience with them has also been that it gets messy (lots of pointer handling, and then a macro every time for retrieving the "next" structure - which makes sense if you're to deal with variable sized headers such as IP headers, or even ethernet headers etc where the "next" structure isn't guaranteed to be what you expect...

So I'm in agreement with the current strategy, but it may need some adjustment at some point in the future - so just be aware of it.  I cannot envision that this will change.